File: sem_2/doubly_linked_list.c
//Author: Sanskriti Gupta
//Date: 5 March 2022
//Aim: To implement doubly linked list 
#include<stdio.h>
#include<stdlib.h>
struct Node
{
    int data;
    struct Node *next,*prev;
};
struct Node *start=NULL;
void insert_beg(int item);
void insert_pos(int item, int pos);
void insert_end(int item);
int delete_beg();
int delete_pos(int pos);
int delete_end();
int linear_search(int item);
void display();
struct Node * create_node();
int main()
{
    int item,choice,pos,del;
    while(1)
    {
        printf("\n1. To Insert at beginning ");
        printf("\n2. To Insert at position ");
        printf("\n3. To Insert at end ");
        printf("\n4. To Delete at beginning ");
        printf("\n5. To Delete at position ");
        printf("\n6. To Delete at end ");
        printf("\n7. To Display ");
        printf("\n8. Linear Search");
        printf("\n9. To Exit \n");
        scanf("%d",&choice);
        switch(choice)
        {
            case 1:
                printf("\nEnter the element: ");
                scanf("%d",&item);
                insert_beg(item);
                break;
            case 2:
                printf("\nEnter the element: ");
                scanf("%d",&item);
                printf("\nEnter the position: ");
                scanf("%d",&pos);
                insert_pos(item,pos);
                break;
            case 3:
                printf("\nEnter the element: ");
                scanf("%d",&item);
                insert_end(item);
                break;
            case 4:
                del=delete_beg();
                printf("\nDeleted Item: %d",del);
                break;
            case 5:
                printf("\nEnter the position: ");
                scanf("%d",&pos);
                del=delete_pos(pos);
                printf("\nDeleted item: %d",del);
                break;
            case 6:
                del=delete_end();
                printf("\nDeleted Item: %d",del);
                break;
            case 7:
                display();
                break;
            case 8:
                printf("\nEnter element to be searched: ");
                scanf("%d",&item);
                int i=linear_search(item);
                if(i)
                {
                    printf("\n%d Found",item);
                }
                else
                {
                    printf("%d NOT found",item);
                }
                break;
            case 9:
                exit(0);
            default:
                printf("\nWrong Choice");
        }
    }
}
struct Node *create_node()
{
    struct Node *temp=(struct Node *)malloc(sizeof(struct Node));
    return temp;
}
void insert_beg(int item)
{
    struct Node *new=create_node();
    new->data=item;
    if(start==NULL)
    {
        new->prev=start;
        start=new;
        new->next=NULL;
    }
    else
    {
        new->prev=start;
        new->next=start;
        start=new;
    }
}
void insert_pos(int item, int pos)
{
    struct Node *new=create_node();
    struct Node *ptr=start;
    int p=1;
    new->data=item;
    if(start==NULL)
    {
        new->next=NULL;
        new->prev=start;
        start=new;
    }
    else
    {
        while(p<pos-1)    
        {
            ptr=ptr->next;
            p++;
        }
        new->next=ptr->next;
        ptr->next->prev=new;
        ptr->next=new;
        ptr=new;
    }
}
void insert_end(int item)
{
    struct Node *new=create_node();
    struct Node *ptr=start;
    new->data=item;
    if(start==NULL)
    {
        new->prev=NULL;
        new->next=NULL;
        start=new;
    }
    else
    {
        while(ptr->next!=NULL)
        {
            ptr=ptr->next;
        }
        ptr->next=new;
        new->prev=ptr;
        new->next=NULL;
    }
}
int delete_beg()
{
    int i;
    if(start==NULL)
    {
        printf("\nEmpty List");
        return -1;
    }
    else if(start->next==NULL)
    {
        i=start->data;
        struct Node *temp=start;
        start=NULL;
        free(temp);
    }
    else
    {
        i=start->data;
        start->next->prev=start;
        start=start->next;
    }
    return i;
}
int delete_end()
{
    int i;
    struct Node *ptr=start;
    if(start==NULL)
    {
        printf("\nEmpty List");
        return -1;
    }
    else if(start->next==NULL)
    {
        i=start->data;
        struct Node *temp=start;
        start=NULL;
        free(temp);
    }
    else
    {
        while(ptr->next!=NULL)
        {
            ptr=ptr->next;
        }
        i=ptr->data;
        ptr->prev->next=NULL;
        free(ptr);
    }
    return i;
}
int delete_pos(int pos)
{
    struct Node *ptr=start;
    int i,p=1;
    if(start==NULL)
    {
        printf("\nList is Empty");
    }
    else
    {
        while(p<pos)
        {
            ptr=ptr->next;
            p++;
        }
        i=ptr->data;
        struct node *temp=ptr;
        ptr->prev->next=ptr->next;
        ptr->next->prev=ptr->prev;
        free(temp);
        return i;
    }
}
int linear_search(int item)
{
    int flag=0;
    if(start->data==item)
    {
        flag++;
    }
    else
    {
        struct Node *ptr=start;
        while(ptr->next!=NULL)
        {
            ptr=ptr->next;
            if(ptr->data==item)
            {
                flag++;
                break;
            }
        }
    }
    return flag;
}
void display()
{
    struct Node *ptr=start;
    while(ptr!=NULL)
    {
        printf("%d ",ptr->data);
        ptr=ptr->next;
    }
}

File: practice/doubly_linked_list.java
import java.util.Scanner;
import java.lang.System;
public class doubly_linked_list 
{
    public static class Node
    {
        int data;
        public Node next;
        public Node prev;
        Node(int data)
        {
            this.data = data;
            this.next = null;
            this.prev = null;
        }
    }
    public static Node start;
    public static Node temp;
    static class doubly_ll_functions
    {
        void ins_beg(int data)
        {
            Node n = new Node(data);
            n.prev = start;
            if(start==null)
            {
                start = n;
            }
            else
            {
                n.next = start;
                start = n;
            }
        }
        void display()
        {
            if(start==null)
            {
                System.out.println("Empty list !!!");
            }
            else
            {
                temp = start;
                while(temp!=null)
                {
                    System.out.print(temp.data + " ");
                    temp = temp.next;
                }
            }
        }
        void display_reverse()
        {
            if(start==null)
            {
                System.out.println("Empty list !!!");
            }
            else
            {
                temp = start;
                while(temp.next!=null)
                {
                    temp = temp.next;
                }
                while(temp != start)
                {
                    System.out.print(temp.data + " ");
                    temp = temp.prev;
                }
            }
        }
    }
    public static void main(String[] args)
    {
        int choice,data;
        Scanner sc = new Scanner(System.in);
        doubly_ll_functions dll = new doubly_ll_functions();
        while(true)
        {
            System.out.println("\n1.Insert at beginning");
            System.out.println("2.Insert at end");
            System.out.println("3.Insert at position");
            System.out.println("4.Delete at beginning");
            System.out.println("5.Delete at end");
            System.out.println("6.Delete at position");
            System.out.println("7.Display in forward");
            System.out.println("8.Display in reverse");
            System.out.println("9.Exit");
            choice = sc.nextInt();
            switch(choice)
            {
                case 1:
                {
                    System.out.print("\nEnter the element: ");
                    data = sc.nextInt();
                    dll.ins_beg(data);
                    break;
                }
                case 7:
                {
                    dll.display();
                    break;
                }
                case 8:
                {
                    dll.display_reverse();
                    break;
                }
                case 9:
                {
                    System.exit(0);
                }
            }
        }
    }   
}